src/encoded/static/components/facets/defaults.js

```diff
@@ -0,0 +1,928 @@
+    filteredTerms: PropTypes.array.isRequired,
+    /** Special search-result click handler */
+    onFilter: PropTypes.func,
+};
+
+FacetTerms.defaultProps = {
+    mode: '',
+    queryString: '',
+    onFilter: null,
+};
+
+
+/**
+ * Display the default text facet with optional typeahead field.
+ */
+export const DefaultFacet = ({ facet, results, mode, relevantFilters, pathname, queryString, onFilter }) => {
+    const [initialState, setInitialState] = React.useState(true);
+    const [topShadingVisible, setTopShadingVisible] = React.useState(false);
+    const [bottomShadingVisible, setBottomShadingVisible] = React.useState(false);
+    const [typeaheadTerm, setTypeaheadTerm] = React.useState('');
+    const scrollingElement = React.useRef(null);
+
+    // Retrieve reference to the registered facet title component for this facet.
+    const TitleComponent = FacetRegistry.Title.lookup(facet.field);
+
+    // Filter out terms with a zero doc_count, as seen in region-search results.
+    const significantTerms = facet.appended !== 'true' ? facet.terms.filter(term => term.doc_count > 0) : facet.terms;
+
+    // Sort numerical terms by value not by frequency
+    // This should ultimately be accomplished in the back end, but the front end fix is much simpler so we are starting with that
+    // We have to check the full list for now (until schema change) because some lists contain both numerical and string terms ('Encyclopedia version' under Annotations) and we do not want to sort those by value
+    const numericalTest = a => !isNaN(a.key);
+    // For straightforward numerical facets, just sort by value
+    const processedTerms = significantTerms.every(numericalTest) ? _.sortBy(significantTerms, obj => obj.key) : significantTerms;
+
+    // Filter the list of facet terms to those allowed by the optional typeahead field. Memoize the
+    // resulting list to avoid needlessly rerendering the facet-term list that can get very long.
+    const filteredTerms = React.useMemo(() => {
+        if (facet.type === 'typeahead') {
+            // Typeahead facets only render a truncated list of terms until initialState becomes
+            // false, which happens when the component mounts.
+            const truncatedProcessedTerms = initialState ? processedTerms.slice(0, 50) : processedTerms;
+            return (
+                truncatedProcessedTerms.filter(
+                    (term) => {
+                        if (term.doc_count > 0) {
+                            const termKey = sanitizedString(term.key);
+                            const typeaheadVal = String(sanitizedString(typeaheadTerm));
+                            if (termKey.match(typeaheadVal)) {
+                                return term;
+                            }
+                            return null;
+                        }
+                        return null;
+                    }
+                )
+            );
+        }
+        return processedTerms;
```

Comment:
It looks like you're doing operations in a different order than the original typeahead code. The original typeahead code filtered to make sure that terms matched searches and then took the first 50 terms if the component hadn't mounted yet. I think by doing the operation in this order, you could end up with an odd situation. Hypothetically, suppose that the first 50 terms for the facet begin with "a" and the user searches for "z". Then, `truncatedProcessedTerms` will get filtered to no terms even though you're trying to get the first 50 terms that match a given search. Is there a reason that you switched the order of operations here?

Also, I'm just curious, the original facet implementation had typeahead facets as a default facet type. Does it work better to have typeahead facets be a type of regular facets? Just wondering if having them separate would make things like this easier, or if this is simpler.

Reply:
I’ll look at the order.

If I understand right, you’re suggesting splitting the typeahead facet renderer out of `DefaultFacet`? Since whether a facet is typeahead or not is determined not by facet.field but by the schema, it made sense to me that typeahead should be part of `DefaultFacet`. I feel like I could be misunderstanding you though.

Reply:
I think I wasn't being clear. I was just sort of wondering why you decided not to have a `DefaultTypeaheadFacet` the way that there's a `DefaultExistsFacet` and a `DefaultDateSelectorFacet`. Maybe it's not too different from a regular facet so it's easier to fold it into `DefaultFacet`?

Reply:
Oh, I see. Yeah, this way if wranglers/associates change a facet from non-typeahead to typeahead, we don’t have to change the corresponding Javascript — DefaultFacet adapts, and their code is largely the same anyway.

Reply:
Regarding the order, I thought the way I had it meant the same thing, but it clearly doesn't. I corrected the order.

Reply:
Regarding the order, I thought the way I had it meant the same thing, but it clearly doesn't. I corrected the order.

_Odd, when I typed this the first time, it showed as a pending review comment so I deleted it. I’ll try it again._

Reply:
Looks good now!